File: firebase/functions/lib/templates/on_template.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../utils/infra/firestore_event_function.dart';
import '../utils/infra/on_firestore_helper.dart';
import '../on_firestore_function.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/templates/template.dart';

import '../utils/template_utils.dart';

class OnTemplate extends OnFirestoreFunction<Template> {
  OnTemplate()
      : super(
          [
            AppFirestoreFunctionData(
              'TemplateOnCreate',
              FirestoreEventType.onCreate,
            ),
          ],
          (snapshot) {
            return TemplateUtils.templateFromSnapshot(snapshot)
                .copyWith(id: snapshot.documentID);
          },
        );

  @override
  String get documentPath => 'community/{communityId}/templates/{templateId}';

  @override
  Future<void> onUpdate(
    Change<DocumentSnapshot> changes,
    Template before,
    Template after,
    DateTime updateTime,
    EventContext context,
  ) async {
    throw UnimplementedError();
  }

  @override
  Future<void> onCreate(
    DocumentSnapshot documentSnapshot,
    Template parsedData,
    DateTime updateTime,
    EventContext context,
  ) async {
    print('Template (${documentSnapshot.documentID}) has been created');

    final communityId = context.params[FirestoreHelper.kCommunityId];
    if (communityId == null) {
      throw ArgumentError.notNull('communityId');
    }

    await onboardingStepsHelper.updateOnboardingSteps(
      communityId,
      documentSnapshot,
      firestoreHelper,
      OnboardingStep.createGuide,
    );
  }

  @override
  Future<void> onDelete(
    DocumentSnapshot documentSnapshot,
    Template parsedData,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }

  @override
  Future<void> onWrite(
    Change<DocumentSnapshot> changes,
    Template before,
    Template after,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }
}

File: firebase/functions/src/templates/on_template.ts
import * as admin from 'firebase-admin';
import * as functions from 'firebase-functions';
import { OnFirestoreFunction, FirestoreEventType } from '../on_firestore_function';
import { firestoreUtils } from '../utils/infra/firestore_utils';
import { onboardingStepsHelper } from '../utils/onboarding_steps_helper';
import { firestoreHelper, FirestoreHelper } from '../utils/infra/on_firestore_helper';
import { TemplateUtils } from '../utils/template_utils';
import { OnboardingStep } from '../types';

interface Template {
  id?: string;
  [key: string]: unknown;
}

type DocumentSnapshot = admin.firestore.DocumentSnapshot;

export class OnTemplate extends OnFirestoreFunction<Template> {
  constructor() {
    super(
      [{ functionName: 'TemplateOnCreate', firestoreEventType: FirestoreEventType.onCreate }],
      (snapshot: DocumentSnapshot) => ({
        ...TemplateUtils.templateFromSnapshot(snapshot),
        id: snapshot.id,
      } as unknown as Template)
    );
  }

  get documentPath(): string {
    return 'community/{communityId}/templates/{templateId}';
  }

  async onCreate(
    documentSnapshot: DocumentSnapshot,
    _parsedData: Template,
    _updateTime: Date,
    context: functions.EventContext
  ): Promise<void> {
    console.log(`Template (${documentSnapshot.id}) has been created`);
    const communityId = context.params[FirestoreHelper.kCommunityId];
    if (!communityId) throw new Error('communityId is null');

    await onboardingStepsHelper.updateOnboardingSteps(
      communityId,
      documentSnapshot,
      firestoreHelper,
      OnboardingStep.createTemplate
    );
  }
}
